File: src/main/java/ch/myniva/gradle/caching/s3/internal/AwsS3BuildCacheService.java
/*
 * Copyright 2017 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.myniva.gradle.caching.s3.internal;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.StorageClass;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import org.gradle.caching.BuildCacheEntryReader;
import org.gradle.caching.BuildCacheEntryWriter;
import org.gradle.caching.BuildCacheException;
import org.gradle.caching.BuildCacheKey;
import org.gradle.caching.BuildCacheService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AwsS3BuildCacheService implements BuildCacheService {

  private static final String BUILD_CACHE_CONTENT_TYPE = "application/vnd.gradle.build-cache-artifact";
  private static final Logger logger = LoggerFactory.getLogger(AwsS3BuildCacheService.class);

  private final AmazonS3 s3;
  private final String bucketName;
  private final String path;
  private final boolean reducedRedundancy;

  AwsS3BuildCacheService(AmazonS3 s3, String bucketName, String path, boolean reducedRedundancy) {
    this.s3 = s3;
    this.bucketName = bucketName;
    this.path = path;
    this.reducedRedundancy = reducedRedundancy;
  }

  private String getBucketPath(BuildCacheKey key) {
    if (path == null || path.length() == 0) {
      return key.getHashCode();
    }
    return (path+"/"+key.getHashCode()).replaceAll("[/]+", "/");
  }

  @Override
  public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) {
    final String bucketPath = getBucketPath(key);
    if (s3.doesObjectExist(bucketName, bucketPath)) {
      logger.info("Found cache item '{}' in S3 bucket", bucketPath);
      S3Object object = s3.getObject(bucketName, bucketPath);
      try (InputStream is = object.getObjectContent()) {
        reader.readFrom(is);
        return true;
      } catch (IOException e) {
        throw new BuildCacheException("Error while reading cache object from S3 bucket", e);
      }
    } else {
      logger.info("Did not find cache item '{}' in S3 bucket", bucketPath);
      return false;
    }
  }

  @Override
  public void store(BuildCacheKey key, BuildCacheEntryWriter writer) {
    final String bucketPath = getBucketPath(key);
    logger.info("Start storing cache entry '{}' in S3 bucket", bucketPath);
    ObjectMetadata meta = new ObjectMetadata();
    meta.setContentType(BUILD_CACHE_CONTENT_TYPE);

    File tempOutputFile;
    try {
      tempOutputFile = File.createTempFile("gradle-s3-build-cache", ".tmp");
    } catch (IOException e) {
      throw new BuildCacheException("Error writing temp file", e);
    }

    try (FileOutputStream os = new FileOutputStream(tempOutputFile)) {
      writer.writeTo(os);
      meta.setContentLength(tempOutputFile.length());
      try (InputStream is = new FileInputStream(tempOutputFile)) {
          PutObjectRequest request = getPutObjectRequest(bucketPath, meta, is);
          if(this.reducedRedundancy) {
            request.withStorageClass(StorageClass.ReducedRedundancy);
          }
          s3.putObject(request);
      }
    } catch (IOException e) {
      throw new BuildCacheException("Error while storing cache object in S3 bucket", e);
    } finally {
      tempOutputFile.delete();
    }
  }

  protected PutObjectRequest getPutObjectRequest(String bucketPath, ObjectMetadata meta, InputStream is) {
    return new PutObjectRequest(bucketName, bucketPath, is, meta);
  }

  @Override
  public void close() throws IOException {
    // The AWS S3 client does not need to be closed
  }
}

File: src/main/kotlin/ch/myniva/gradle/caching/s3/internal/AwsS3BuildCacheService.kt
/*
 * Copyright 2017 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ch.myniva.gradle.caching.s3.internal

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.AmazonS3Exception
import com.amazonaws.services.s3.model.ObjectMetadata
import com.amazonaws.services.s3.model.PutObjectRequest
import com.amazonaws.services.s3.model.StorageClass
import org.gradle.api.logging.LogLevel
import org.gradle.api.logging.Logging
import org.gradle.caching.*
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException

private val logger = Logging.getLogger(AwsS3BuildCacheService::class.java)

class AwsS3BuildCacheService internal constructor(
    private val s3: AmazonS3,
    private val bucketName: String,
    private val path: String?,
    private val reducedRedundancy: Boolean,
    private val maximumCachedObjectLength: Long
) : BuildCacheService {
    companion object {
        private const val BUILD_CACHE_CONTENT_TYPE = "application/vnd.gradle.build-cache-artifact"
    }

    private val cacheLoads = Stopwatch()
    private val cacheHits = Stopwatch()
    private val cacheStores = Stopwatch()

    override fun close() {
        fun Long.mib() = (this + 512L * 1024) / (1024L * 1024)

        s3.shutdown()
        if (cacheLoads.starts != 0) {
            logger.log(
                if (cacheHits.elapsed > 1000) LogLevel.LIFECYCLE else LogLevel.INFO,
                "S3 cache reads: ${cacheHits.starts}, requests: ${cacheLoads.starts}, " +
                        "elapsed time: ${cacheLoads.elapsed}ms, processed: ${cacheLoads.bytes.mib()}MiB"
            )
        }
        if (cacheStores.starts != 0) {
            logger.lifecycle(
                "S3 cache writes: ${cacheStores.starts}, " +
                        "elapsed time: ${cacheStores.elapsed}ms, sent to cache: ${cacheStores.bytes.mib()}MiB"
            )
        }
    }

    private fun BuildCacheKey.getBucketPath() = if (path.isNullOrEmpty()) {
        hashCode
    } else {
        "$path/$hashCode"
    }

    override fun load(key: BuildCacheKey, reader: BuildCacheEntryReader): Boolean = cacheLoads {
        loadInternal(key, reader)
    }

    private fun Stopwatch.loadInternal(key: BuildCacheKey, reader: BuildCacheEntryReader): Boolean {
        val bucketPath = key.getBucketPath()
        try {
            s3.getObject(bucketName, bucketPath).use { s3Object ->
                val contentLength = s3Object.objectMetadata.contentLength
                if (contentLength > maximumCachedObjectLength) {
                    logger.info(
                        "Cache item '{}' '{}' in S3 bucket size is {}, and it exceeds maximumCachedObjectLength {}. Will skip the retrieval",
                        key.displayName,
                        bucketPath,
                        contentLength,
                        maximumCachedObjectLength
                    )
                    return false
                }
                bytesProcessed(contentLength)
                cacheHits {
                    reader.readFrom(s3Object.objectContent)
                }
            }
            return true
        } catch (e: AmazonS3Exception) {
            if (e.statusCode == 403 || e.statusCode == 404) {
                logger.info(
                    when (e.statusCode) {
                        403 -> "Got 403 (Forbidden) when fetching cache item '{}' '{}' in S3 bucket"
                        else -> "Did not find cache item '{}' '{}' in S3 bucket"
                    },
                    key.displayName,
                    bucketPath
                )
                return false
            }
            logger.info(
                "Unexpected error when fetching cache item '{}' '{}' in S3 bucket",
                key.displayName,
                bucketPath,
                e
            )
            return false;
        }
    }

    override fun store(key: BuildCacheKey, writer: BuildCacheEntryWriter) = cacheStores {
        storeInternal(key, writer)
    }

    private fun Stopwatch.storeInternal(key: BuildCacheKey, writer: BuildCacheEntryWriter) {
        val bucketPath = key.getBucketPath()
        val itemSize = writer.size
        if (itemSize > maximumCachedObjectLength) {
            logger.info(
                "Cache item '{}' '{}' in S3 bucket size is {}, and it exceeds maximumCachedObjectLength {}. Will skip caching it.",
                key.displayName,
                bucketPath,
                itemSize,
                maximumCachedObjectLength
            )
            return
        }
        bytesProcessed(itemSize)
        logger.info("Start storing cache entry '{}' in S3 bucket", bucketPath)
        val meta = ObjectMetadata().apply {
            contentType = BUILD_CACHE_CONTENT_TYPE
            contentLength = writer.size
        }
        try {
            val stream = ByteArrayInputStream(
                ByteArrayOutputStream()
                    .also { os -> writer.writeTo(os) }
                    .toByteArray()
            )
            val request = PutObjectRequest(bucketName, bucketPath, stream, meta)
            if (reducedRedundancy) {
                request.withStorageClass(StorageClass.ReducedRedundancy)
            }
            s3.putObject(request)
        } catch (e: IOException) {
            throw BuildCacheException("Error while storing cache object in S3 bucket", e)
        }
    }
}
